core/src/main/java/org/elasticsearch/cluster/service/ClusterService.java

```diff
@@ -701,152 +725,219 @@ public String source() {
-                } catch (Exception ex) {
-                    logger.warn("failed to notify ClusterStateListener", ex);
-                }
-            }
+        nodeConnectionsService.disconnectFromNodes(clusterChangedEvent.nodesDelta().removedNodes());
 
-            //manual ack only from the master at the end of the publish
-            if (newClusterState.nodes().isLocalNodeElectedMaster()) {
-                try {
-                    ackListener.onNodeAck(newClusterState.nodes().getLocalNode(), null);
-                } catch (Exception e) {
-                    final DiscoveryNode localNode = newClusterState.nodes().getLocalNode();
-                    logger.debug(
-                        (Supplier<?>) () -> new ParameterizedMessage("error while processing ack for master node [{}]", localNode),
-                        e);
-                }
-            }
+        updateState(css -> new ClusterServiceState(css.getClusterState(), ClusterStateStatus.APPLIED));
 
-            for (UpdateTask<T> task : proccessedListeners) {
-                task.listener.clusterStateProcessed(task.source, previousClusterState, newClusterState);
+        for (ClusterStateListener listener : postAppliedListeners) {
+            try {
+                logger.trace("calling [{}] with change to version [{}]", listener, newClusterState.version());
+                listener.clusterChanged(clusterChangedEvent);
+            } catch (Exception ex) {
+                logger.warn("failed to notify ClusterStateListener", ex);
             }
+        }
 
+        //manual ack only from the master at the end of the publish
+        if (newClusterState.nodes().isLocalNodeElectedMaster()) {
             try {
-                executor.clusterStatePublished(clusterChangedEvent);
+                ackListener.onNodeAck(newClusterState.nodes().getLocalNode(), null);
             } catch (Exception e) {
-                logger.error(
-                    (Supplier<?>) () -> new ParameterizedMessage(
-                        "exception thrown while notifying executor of new cluster state publication [{}]",
-                        tasksSummary),
+                final DiscoveryNode localNode = newClusterState.nodes().getLocalNode();
+                logger.debug(
+                    (Supplier<?>) () -> new ParameterizedMessage("error while processing ack for master node [{}]", localNode),
                     e);
             }
+        }
 
-            TimeValue executionTime = TimeValue.timeValueMillis(Math.max(0, TimeValue.nsecToMSec(currentTimeInNanos() - startTimeNS)));
-            logger.debug("processing [{}]: took [{}] done applying updated cluster_state (version: {}, uuid: {})", tasksSummary,
-                executionTime, newClusterState.version(), newClusterState.stateUUID());
-            warnAboutSlowTaskIfNeeded(executionTime, tasksSummary);
+        taskOutputs.processedDifferentClusterState(previousClusterState, newClusterState);
+
```

Comment:
there is one newline before it and one after - This is deliberate so that it stands out as a single step in the method.